CPU offloading for GPUs: generate memcpy D2H/H2D

zhenying-liu opened the pull request:
This memory offload implementation includes:
- Add the mlir for copy-start/copy-done
- Generate copy_thunk from HLO instructions for the new runtime
- Add the host memory allocator/deallocator for the host memory space
- Generate async memcpy d2h/h2d

Tested locally with a Jax code, by 
- enabling the HLO rematerialization offloading to create copy-start/copy-done
- disabling recompute/compress 

The log showed memcpy d2h and h2d were created by copy_thunk, meaning this offload implementation works.

sergey-kozub reviewed (approved):
CompiledProgramsCount test seems to not be related to the changes in this PR (should be in a separate PR?), but otherwise looking good.

zhenying-liu commented:
@sergey-kozub Thanks for the comments. Will remove CompiledProgramsCount from the unit test.

ezhulenev commented:
FYI we plan to remove LMHLO ASAP (plan start removing it in ~January), can you also add a code path that is compatible with thunks runtime, here is a recent example for adding Send/Recv support (ported from lmhlo): https://github.com/openxla/xla/commit/dbcd8f2f5f673fe7ebb5d2d2321ebda57fc5aa67

The gist is that all thunks should be emitted from `HloInsruction*` instead of `mlir::Operation`.

You can test that it works with `XLA_FLAGS=--xla_gpu_enable_xla_runtime_executable=false`

zhenying-liu commented:
Thanks @ezhulenev. The Send/Recv example is really helpful. And the code is neat! Will add this code path.

nouiz commented:
> CompiledProgramsCount test seems to not be related to the changes in this PR (should be in a separate PR?), but otherwise looking good.

@sergey-kozub 
CompiledProgramsCount  is the real test for this PR. It take an not optimized graph, call the opt pass that add the offloading instruction. Then execute that optimized graph vs the not optimized graph.

What make you think it isn't related to this PR?

kamaljeeti commented:
Hi @zhenying-liu, This PR is in draft, any update on this? Please. Thank you!

zhenying-liu commented:
@kamaljeeti I figured out how to generate memcpy D2H and H2D for copy-start. But I don't have a good solution for the memory deallocation to differentiate the device/host memory deallocation. Also Trevor made 3 PRs: #7843, #7962 and #7963 for NCCL memory allocator. I'm merging with his PRs locally and tries to try to make it work similarly.

nouiz commented:
@cheshire Who can review this?

cheshire commented:
@SandSnip3r presumably, since he's leading the host offloading implementation. I could help as well if there are contentious points.

cheshire reviewed (commented):
Overall looks reasonable to me, but I think Victor should take a look.

SandSnip3r commented on xla/pjrt/gpu/se_gpu_pjrt_client.cc:
```diff
@@ -792,6 +792,14 @@ GetStreamExecutorGpuDeviceAllocator(
+        GetGpuHostAllocator(ordinal_and_device.second->executor());
+    allocators.emplace_back(std::move(host_allocator),
+                            ordinal_and_device.second->compute_stream(),
+                            /*memory_space=*/5);
```
Magic number, does it make sense to name this memory space and put it in a constant?

zhenying-liu replied:
@SandSnip3r Yes, I can put it in se_gpu_pjrt_client.h.  
I put "constexpr int kHostMemorySpace = 5" in runtime3/copy_thunk.h. But I don't have a good solution to just initialize it once in one header file. Do you have any suggestion?

SandSnip3r commented on xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -3966,6 +3966,50 @@ static std::optional<GlobalDeviceId> DeviceConstraint(
+  Shape shape = instr->shape();
+  CHECK(shape.IsTuple());
+
+  MemoryCopyType memory_copy_type = MemoryCopyType::kDeviceToDevice;
```
Maybe instead of the pattern
```
Type t = defaultValue;
if (a) {
  t = 1;
}
if (b) {
  t = 2;
}
TF_RET_CHECK(t != defaultValue);
```
You could not need to rely on a default value which is unused by doing the following:
```
Type t;
if (a) {
  t = 1;
} else if (b) {
  t = 2;
} else {
  return Internal("Good error message");
}
```
That is, assuming the two are mutually exclusive.

SandSnip3r replied:
In fact, I think they're not mutually exclusive, you could have an async copy from host to host. 
Your code will set `memory_copy_type` to `kHostToDevice`.

zhenying-liu replied:
In this context, the two are mutually exclusive because the copy_start can generate host-to-device or device-to-host, no device-to-device. Thanks for showing me the pattern and I'm going to change correspondingly.

SandSnip3r commented on xla/service/gpu/runtime3/copy_thunk.cc:
```diff
@@ -22,29 +22,50 @@ limitations under the License.
 namespace xla {
 namespace gpu {
 
 DeviceToDeviceCopyThunk::DeviceToDeviceCopyThunk(
```
Seems like "DeviceToDeviceCopy" is no longer an appropriate name

zhenying-liu replied:
OK, I will create DeviceToHostCopy/HostToDeviceCopy thunks, leaving DeviceToDeviceCopyThunk alone, as it is used in other places.

SandSnip3r commented on xla/service/gpu/ir_emitter_unnested.cc:
```diff
@@ -4563,6 +4622,7 @@ absl::Status IrEmitterUnnested::EmitHloInstruction(
     case HloOpcode::kGetTupleElement:
     case HloOpcode::kParameter:
     case HloOpcode::kTuple:
+    case HloOpcode::kCopyDone:
```
Does this mean that the copy-done is a no-op? In Hlo, we expect that this op would block until the copy is completed. Is that how this works?

zhenying-liu replied:
Treating copy-done as a no-op seems working. I'll double check with my coworker.

zhenying-liu replied:
I got an nsys report showing that the memcpy H2D/D2H that I generated are in the same stream (Stream #118) of other compute operations like wrapped_add_1. So treating copy-done as a no-op works now. The future PR will use a separate stream for memcpy and the blocking or events will be absolutely necessary when handling copy-done. So far we can leave copy-done as is for the current PR.
<img width="1314" alt="nsys-offloading" src="https://github.com/openxla/xla/assets/90360117/5bbf092e-70a8-493d-8ef6-125fd29ca476">

SandSnip3r commented on xla/service/async_collective_creator.cc:
```diff
@@ -123,6 +123,18 @@ StatusOr<ReplacedAsync> CreateAsyncStartDone(
   return ReplacedAsync{start, done};
 }
 
+StatusOr<ReplacedAsync> CreateCopyStartDone(HloInstruction* instruction) {
```
Copy is not a collective and should not be put into the `AsyncCollectiveCreator`. Either change the pass to be a generic async creator, or create a new pass for the async copy.

zhenying-liu replied:
Sure, it doesn't belong to collectives. Will change as you said.

SandSnip3r reviewed (changes requested):
This PR needs tests.

zhenying-liu commented:
> This PR needs tests.

I previously created an hlo test with copy_start/copy_done, and also generate copy_start/copy_done from HLO rematerialization. But the hlo test doesn't go through PjRT routine where I got the host memory allocator. So I won't have a working unit test unless I pass the host memory allocator which seems not obvious for me. I'll ask for the help.

zhenying-liu commented:
I addressed all the issues that @SandSnip3r asked, except from a unit test. I will follow up with the unit test. Since I couldn't call a multidevice allocator in the hlo test easily, I plan to check if memcpy D2H/H2D are generated for the given HLO IR with copy-start/copy-done. 
Sorry for wiping out the git history as I can't keep the history and sync with the origin/main cleanly.
The major change in this commit is:
- added DeviceToHostCopyThunk and HostToDeviceCopyThunk in copy_thunk.cc/copy_thunk.h to avoid using DeviceToDeviceThunk.

yashk2810 commented:
Can you sync to HEAD so that we can get this PR in?

SandSnip3r commented:
> The CI failure on XLA Linux ARM64 CPU wasn't introduce by me.

That's right. The breakage is known internally and we're asking for a fix. I'll let you know when that gets fixed and then you can re-sync.

SandSnip3r commented:
I think https://github.com/openxla/xla/pull/9675 fixed it. Try syncing.

zhenying-liu commented:
> I think #9675 fixed it. Try syncing.

Thanks Victor. I synced again and now CI build is fine. Please take a look.

zhenying-liu commented:
@SandSnip3r I got this Google internal checks FAILED for runs and I don't have access to details. Those are related with removing the last two mlir operands from copy thunk: 
`mlir::Value source_value, mlir::Value destination_value`
I guess maybe Google internal tests still assume the existence of these two mlir operands. Please help check and provide error messages or information so that I can work on it. Thanks!

zhenying-liu commented on xla/service/gpu/BUILD:
```diff
@@ -3645,6 +3646,31 @@ xla_test(
+    srcs = ["gpu_offloading_test.cc"],
+    tags = tf_cuda_tests_tags(),
+    deps = [
+        ":horizontal_loop_fusion",
```
Thanks. I fixed in the latest commit.
